src/config.rs

```diff
@@ -2,16 +2,118 @@ pub mod settings;
 pub mod storage;
 
 use crate::config::settings::{Setting, SettingInfo};
-use crate::types::Result;
+use crate::config::storage::memory::MemoryStorageAdapter;
+use lazy_static::lazy_static;
+use log::warn;
 use serde_derive::Deserialize;
 use serde_json::Value;
 use std::collections::HashMap;
 use std::mem;
 use std::str::FromStr;
+use std::sync::RwLock;
 use wildmatch::WildMatch;
 
+/// Settings are stored in a json file, but this is included in the binary for mostly easy editting.
 const SETTINGS_JSON: &str = include_str!("./config/settings.json");
 
+/// StoreAdapter is the interface for storing and retrieving settings
+/// This can be used to storage settings in a database, json file, etc
+/// Note that we need to implement Send so we can send the storage adapter
+/// to other threads.
+pub trait StorageAdapter: Send + Sync {
+    /// Retrieves a setting from the storage
+    fn get(&self, key: &str) -> Option<Setting>;
+
+    /// Stores a given setting to the storage. Note that "self" is self and not "mut self". We need to be able
+    /// to storage settings in a non-mutable way. That is mostly possible it seems with a mutex lock that we
+    /// can get mutable.
+    fn set(&self, key: &str, value: Setting);
+
+    /// Retrieves all the settings in the storage in one go. This is used for preloading the settings
+    /// into the ConfigStore and is more performant normally than calling get_setting manually for each
+    /// setting.
+    fn all(&self) -> crate::types::Result<HashMap<String, Setting>>;
+}
+
+lazy_static! {
+    // Initial config store will have a memory storage adapter. It will save within the session, but not
+    // persist this on disk.
+    static ref CONFIG_STORE: RwLock<ConfigStore> = RwLock::new(ConfigStore::default());
```

Comment:
RwLock since we can have multiple reads (for instance, https://github.com/gosub-browser/gosub-engine/pull/267/files#diff-d55130faaf1118028faf388a306896363c166c769c0bab6bdff2b23051659575R101)